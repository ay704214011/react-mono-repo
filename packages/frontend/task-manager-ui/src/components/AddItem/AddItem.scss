button {
    height: 35px;
    width: 100px;
    border: solid 1px blue;
    cursor: pointer;
}

input[type="text"] {
    height: 35px;
    width: 150px;
    border: solid 1px blue;
}