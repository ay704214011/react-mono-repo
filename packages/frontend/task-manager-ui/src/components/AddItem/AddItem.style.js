import { css } from 'styled-components';

export default css`
    .grid-container {
      display: grid;
      grid-template-columns: auto auto auto;
      grid-template-rows: 100px 100px;
      grid-row-gap: 5px;
      grid-column-gap: 5px
    }
    .grid-item-full-width {
       grid-column-start: 1;
       grid-column-end: 4;
    }
    .form-contanier {
       display: grid;
       grid-template-columns: auto auto;
       grid-template-rows: repeat(2, 1fr) 2fr;
       grid-row-gap: 5px;
       grid-column-gap: 5px;

       input, textarea {
         width: 250px;
         height: 30px;
         border: solid 2px grey;
       }
    }
    .form-item {
       text-align: left;
    }
`;