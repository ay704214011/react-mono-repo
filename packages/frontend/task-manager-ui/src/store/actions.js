export const FETCH_LABELS = 'ITEMS/FETCH_LABELS';
export const FETCH_LABELS_SUCCESS = 'ITEMS/FETCH_LABELS_SUCCESS';
export const FETCH_LABELS_FAILURE = 'ITEMS/FETCH_LABELS_FAILURE';
export const ADD_ITEM = 'ITEMS/ADD_ITEM';
export const ADD_ITEMS = 'ITEMS/ADD_ITEMS';
export const DELETE_ITEM = 'ITEMS/DELETE_ITEM';

//Get items actions
export const FETCH_ITEMS = 'ITEMS/FETCH_ITEMS';
export const FETCH_ITEMS_SUCCESS = 'ITEMS/FETCH_ITEMS_SUCCESS';
export const FETCH_ITEMS_FAILURE = 'ITEMS/FETCH_ITEMS_FAILURE';

export const addItem = (item) => ({
  type: ADD_ITEM,
  data: item
});

export const addItems = (items) => ({
    type: ADD_ITEMS,
    data: items
});

export const deleteItem = (id) => ({
  type: DELETE_ITEM,
  id
});

export const fetchLabels = () => ({
  type: FETCH_LABELS
});

export const fetchLabelsSuccess = (data) => ({
  type: FETCH_LABELS_SUCCESS,
  data
});

export const fetchLabelsFailure = () => ({
  type: FETCH_LABELS_FAILURE
});

export const fetchItems = () => ({
  type: FETCH_ITEMS
});

export const fetchItemsSuccess = (data) => ({
  type: FETCH_ITEMS_SUCCESS,
  data
});

export const fetchItemsFailure = () => ({
  type: FETCH_ITEMS_FAILURE
});