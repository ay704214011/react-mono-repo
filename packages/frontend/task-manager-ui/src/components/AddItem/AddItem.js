import React, { useState, useContext, useCallback } from 'react';
import withStyle from '../../HOC/withStyle';
import styles from './AddItem.style';
import ItemList from "../ItemList/ItemList";
import { StoreContext } from '../../store/store';
import { addItem } from '../../store/actions';
import Button from '../../UI/Button';
import Input from '../../UI/Input';
//import './AddItem.scss';

const AddItem = ({ className }) => {
    const { state, dispatch } = useContext(StoreContext);
    const [name, setName] = useState('');
    const { content: { labels }} = state;

    const onAddItem = useCallback(() => {
        if (name === '') return;
        dispatch(addItem({
          id: Math.random(),
          name: name ? name.charAt(0).toUpperCase() + name.slice(1) : ''
        }));
        setName('');
      }, [dispatch, name]);
    
    const onReset = () => {
      setName('');
    };

    return (
        <div className={className}>
          <p>
              {labels.title}
          </p>
          <div className="grid-container">
            <div className="grid-item">
               <label>{labels.item}:</label>
               <Input className="ml-8" type="text" value={name} onChange={(event) => {
                  setName(event.target.value);
               }}/>
            </div>
            <div className="grid-item">
               <Button onClick={(onAddItem)} className="ml-8" primary data-testid="btn-add-item">{labels.addItem}</Button>
            </div>
            <div className="grid-item">
               <Button onClick={(onReset)} className="ml-8">{labels.reset}</Button>
            </div>
            <div className="grid-item grid-item-full-width">
              <ItemList items={state.items.itemList} dispatch={dispatch} state={state}/>
            </div>
          </div>
          <div className="form-contanier">
            <div className="form-item">
              <label for="">Name</label>
            </div>
            <div className="form-item">
              <input id="name" className="ml-8" type="text"></input>
            </div>
            <div className="form-item">
               <label for="">Category</label>
            </div>
            <div className="form-item">
              <input id="category" className="ml-8" type="text"></input>
            </div>
            <div className="form-item">
               <label for="">Description</label>
            </div>
            <div className="form-item">
              <textarea id="description" className="ml-8" rows="30" cols="50"></textarea>
            </div>
            <div className="form-item">
               <Button className="ml-8" primary>{labels.addItem}</Button>
            </div>
            <div className="form-item">
               <Button className="ml-8">{labels.reset}</Button>
            </div>
          </div>
        </div>
    );
};

export default withStyle(AddItem, styles);