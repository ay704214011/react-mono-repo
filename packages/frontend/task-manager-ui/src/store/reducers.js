import { ADD_ITEM, ADD_ITEMS, DELETE_ITEM, FETCH_LABELS, FETCH_LABELS_SUCCESS, FETCH_LABELS_FAILURE, FETCH_ITEMS, FETCH_ITEMS_SUCCESS, FETCH_ITEMS_FAILURE } from './actions';

export const itemsReducer = (state, action) => {
   const { data } = action;
   switch (action.type) {
      case ADD_ITEM: {
         return {
            ...state,
            itemList: [
               ...state.itemList,
               data
           ]
         }
      }
      case ADD_ITEMS: {
         return {
            ...state,
            itemList: data.itemList
         }
      }
      case DELETE_ITEM: {
        const updatedItems = state.itemList.filter((item) => {
           return item.id !== action.id;
        });
        return {
            ...state,
            itemList: updatedItems
        };
      }
      case FETCH_ITEMS: {
         return {
            ...state,
            isFetchingItems: true
         };
      }
      case FETCH_ITEMS_SUCCESS: {
         return {
            ...state,
            isFetchingItems: false,
            fetchedItems: data
         };
      }
      case FETCH_ITEMS_FAILURE: {
         return {
            ...state,
            isFetchingItems: false,
            isFetchItemsError: true
         };
      }
      default:
        return {
          ...state
        };
   }
};

export const contentReducer = (state, action) => {
   console.log('state 1', state);
   switch (action.type)  {
       case FETCH_LABELS:
         return {
            ...state,
            isFetchingLabels: true
         };
      case FETCH_LABELS_SUCCESS:
         return {
            ...state,
            labels: action.data,
            isFetchingLabels: false
         };
      case FETCH_LABELS_FAILURE:
         return {
           ...state,
           isFetchingLabels: false
         };
      default:
         return {
            ...state
         };
   }
};

export default itemsReducer;