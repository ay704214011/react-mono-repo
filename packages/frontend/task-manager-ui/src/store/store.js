import { createContext } from "react";
import combineReducers from "react-combine-reducers";
import { itemsReducer, contentReducer } from './reducers';

const contentState = {
   isFetchingLabels: false,
   labels: {}
};
const itemsState = {
   itemList: []
};
export const store = {
   items: itemsState,
   content: contentState
};

const [combinedReducer, combinedState] = combineReducers({
  items: [itemsReducer, itemsState],
  content: [contentReducer, contentState]
});

export const StoreContext = createContext({});
export const storeReducer = combinedReducer;
export default combinedState;