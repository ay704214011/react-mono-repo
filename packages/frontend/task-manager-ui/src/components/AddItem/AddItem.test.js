import React from 'react';
import ReactDom from 'react-dom/client';
import AddItem from './AddItem';
import { render, screen } from '@testing-library/react';
import { StoreContext } from '../../store/store';
import labels from '../../content/content';
const context = {
    state: {
        content: {
            labels
        },
        items: {
            itemList: []
        }
    },
    dispatch: jest.fn()
};

describe('Add Item component tests', () => {
    it('should render without crash', () => {
        const wrapper = document.createElement('div');
        const root = ReactDom.createRoot(wrapper);
        root.render(<AddItem />);
    });

    it('should render correctly', () => {
        const { container } = render(
            <StoreContext.Provider value={context}>
                <AddItem />
            </StoreContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with passed content', () => {
        render(
            <StoreContext.Provider value={context}>
                <AddItem />
            </StoreContext.Provider>
        );
        expect(screen.getByTestId('btn-add-item')).toHaveTextContent('Add item');
    });
});