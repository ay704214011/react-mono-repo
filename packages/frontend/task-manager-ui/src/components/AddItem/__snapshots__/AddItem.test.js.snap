// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Item component tests should render correctly 1`] = `
.c0 .grid-container {
  display: grid;
  grid-template-columns: auto auto auto;
  grid-template-rows: 100px 100px;
  grid-row-gap: 5px;
  grid-column-gap: 5px;
}

.c0 .grid-item-full-width {
  grid-column-start: 1;
  grid-column-end: 4;
}

.c0 .form-contanier {
  display: grid;
  grid-template-columns: auto auto;
  grid-template-rows: repeat(2,1fr) 2fr;
  grid-row-gap: 5px;
  grid-column-gap: 5px;
}

.c0 .form-contanier input,
.c0 .form-contanier textarea {
  width: 250px;
  height: 30px;
  border: solid 2px grey;
}

.c0 .form-item {
  text-align: left;
}

.c1 {
  height: 35px;
  width: 150px;
  border: solid 1px grey;
}

.c2 {
  height: 35px;
  width: 100px;
  border: solid 1px grey;
  cursor: pointer;
  background-color: #82c823;
  color: #fff;
}

.c3 {
  height: 35px;
  width: 100px;
  border: solid 1px grey;
  cursor: pointer;
  background-color: #fff;
  color: #000;
}

<div>
  <div
    class="c0"
  >
    <p>
      Shopping List
    </p>
    <div
      class="grid-container"
    >
      <div
        class="grid-item"
      >
        <label>
          Item
          :
        </label>
        <input
          class="c1 ml-8"
          type="text"
          value=""
        />
      </div>
      <div
        class="grid-item"
      >
        <button
          class="c2 ml-8"
          data-testid="btn-add-item"
        >
          Add item
        </button>
      </div>
      <div
        class="grid-item"
      >
        <button
          class="c3 ml-8"
        >
          Reset
        </button>
      </div>
      <div
        class="grid-item grid-item-full-width"
      />
    </div>
    <div
      class="form-contanier"
    >
      <div
        class="form-item"
      >
        <label
          for=""
        >
          Name
        </label>
      </div>
      <div
        class="form-item"
      >
        <input
          class="ml-8"
          id="name"
          type="text"
        />
      </div>
      <div
        class="form-item"
      >
        <label
          for=""
        >
          Category
        </label>
      </div>
      <div
        class="form-item"
      >
        <input
          class="ml-8"
          id="category"
          type="text"
        />
      </div>
      <div
        class="form-item"
      >
        <label
          for=""
        >
          Description
        </label>
      </div>
      <div
        class="form-item"
      >
        <textarea
          class="ml-8"
          cols="50"
          id="description"
          rows="30"
        />
      </div>
      <div
        class="form-item"
      >
        <button
          class="c2 ml-8"
        >
          Add item
        </button>
      </div>
      <div
        class="form-item"
      >
        <button
          class="c3 ml-8"
        >
          Reset
        </button>
      </div>
    </div>
  </div>
</div>
`;
